zilliqa/src/exec.rs

```diff
@@ -490,9 +492,57 @@ impl State {
     }
 
     /// Applies a state delta from a Scilla execution to the state.
-    fn apply_delta_scilla(&mut self, state: &HashMap<Address, Account>) -> Result<()> {
+    fn apply_delta_scilla(&mut self, state: &HashMap<Address, PendingAccount>) -> Result<()> {
         for (&address, account) in state {
-            self.save_account(address, account.clone())?;
+            let mut storage = self.get_account_trie(address)?;
+
+            /// Recursively called internal function which assigns `value` at the correct key to `storage`.
+            fn handle(
+                storage: &mut EthTrie<TrieStorage>,
+                var: &str,
+                value: &StorageValue,
+                indices: &mut Vec<Vec<u8>>,
+            ) -> Result<()> {
+                match value {
+                    StorageValue::Map { map, complete } => {
+                        // If this is a complete view of the map, delete any existing values first.
+                        if *complete {
+                            storage.remove_by_prefix(&storage_key(var, indices))?;
+                        }
+
+                        // Add the next level of index before making the recursive call.
+                        indices.push(vec![]);
+                        for (k, v) in map {
+                            indices.last_mut().unwrap().clone_from(k);
+                            handle(storage, var, v, indices)?;
+                        }
+                    }
+                    StorageValue::Value(Some(value)) => {
+                        let key = storage_key(var, indices);
+                        storage.insert(&key, value)?;
+                    }
+                    StorageValue::Value(None) => {
+                        let key = storage_key(var, indices);
+                        // A deletion may occur at any depth of the map. Therefore we remove all keys with the relevent
+                        // prefix.
+                        storage.remove_by_prefix(&key)?;
```

Comment:
The Scilla interpreter has no way of setting `foo[bar] = None`, other than `delete foo[bar]`. Just to be clear `StorageValue::Value(None)` always means a deletion - The values are not actually optional. Of course, you can use `Option`s at the Scilla language level, but both `Some` and `None` on Scilla's side will turn up as `Some(value)` here.

The only reason I have `StorageValue::Value(Option<Bytes>)` rather than `StorageValue::Value(Bytes)` and `StorageValue::Deleted` is so that `load_storage` can return `&mut Option<Bytes>` and the caller can set it to `None` to delete a value.

I will add a comment in `StorageValue` to clarify the usage. I hope my explanation made sense?

Reply:
Yep. Absolutely - and agree that a comment would have been useful to me. Thanks!